### PR TITLE
[FSTORE-1452] Deduplicate job- and execution-related code

### PR DESCRIPTION
**Pull request by aversey**
JIRA Issue: FSTORE-1452

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [ ] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/tests/core/test_feature_view_engine.py [hunk 1]
```diff
@@ -32,50 +33,51 @@
...
-    name="test4",
-    version=1,
-    featurestore_id=99,
-    primary_key=[],
-    partition_key=[],
-    features=[feature.Feature("id4"), feature.Feature("label4")],
-    id=14,
-    stream=False,
-)
+with mock.patch("hopsworks_common.client.get_instance"):
```
Comment (vatj) [inline python/tests/core/test_feature_view_engine.py:36]: Is this relevant to this PR?

File: python/tests/core/test_statistics_engine.py [hunk 2]
```diff
@@ -30,15 +31,16 @@
...
-fg = feature_group.FeatureGroup(
-    name="test",
-    version=1,
-    featurestore_id=99,
-    primary_key=[],
-    partition_key=[],
-    id=10,
-    stream=False,
-)
+with mock.patch("hopsworks_common.client.get_instance"):
```
Comment (vatj) [inline python/tests/core/test_statistics_engine.py:34]: Is this relevant to this PR?

File: python/tests/core/test_transformation_function_engine.py [hunk 3]
```diff
@@ -27,48 +29,49 @@
...
-        feature.Feature("id"),
-        feature.Feature("tf_name"),
-        feature.Feature("tf1_name"),
-        feature.Feature("tf3_name"),
-    ],
-    id=12,
-    stream=False,
-)
-engine.init("python")
+with mock.patch("hopsworks_common.client.get_instance"):
```
Comment (vatj) [inline python/tests/core/test_transformation_function_engine.py:32]: Is this relevant to this PR?

**Review by vatj — CHANGES_REQUESTED** [review]

File: python/hopsworks_common/core/job_api.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+        headers = {"content-type": "application/json"}
+        return execution.Execution.from_response_json(
+            _client._send_request(
+                "GET", path_params, headers=headers, query_params=query_params
+            ),
+            job=job,
+        )
+
+    def create_or_update_schedule_job(
```
[the remark this answers is not in the document]
Comment (aversey) [reply python/hopsworks_common/core/job_api.py:251]: Well, firstly, it looks like job and schedule just should be reversed, so that instead of `create_or_update_schedule_job` and `delete_schedule_job` it will be `create_or_update_job_schedule` and `delete_job_schedule`.

Second, instead of `create_or_update` I think it can be simply `upload` or `set`, resulting in `upload_job_schedule` / `set_job_schedule`.

File: python/hopsworks_common/execution.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    def json(self):
+        return json.dumps(self, cls=util.Encoder)
+
+    def __str__(self):
+        return self.json()
+
+    def __repr__(self):
+        return f"Execution({self._final_status!r}, {self._state!r}, {self._submission_time!r}, {self._args!r})"
+
+    def get_url(self):
```
[the remark this answers is not in the document]
Comment (aversey) [reply python/hopsworks_common/execution.py:249]: I have also changed secret.py to follow the same style.

File: python/tests/core/test_feature_view_engine.py (hunk 1 above)
Comment (aversey) [reply python/tests/core/test_feature_view_engine.py:36]: Yes, since `FeatureGroup.__init__` calls `VariableApi.__init__`.

File: python/tests/constructor/test_query.py [hunk 6]
```diff
@@ -22,53 +23,55 @@
...
-        features=[
-            feature.Feature("id", feature_group_id=13),
-            feature.Feature("tf_name", feature_group_id=13),
-            feature.Feature("tf1_name", feature_group_id=13),
-            feature.Feature("tf3_name", feature_group_id=13),
-        ],
-        id=13,
-        stream=False,
-    )
+    with mock.patch("hopsworks_common.client.get_instance"):
```
[the remark this answers is not in the document]
Comment (aversey) [reply python/tests/constructor/test_query.py:26]: Yes, since `FeatureGroup.__init__` calls `VariableApi.__init__`.

File: python/tests/core/test_statistics_engine.py (hunk 2 above)
Comment (aversey) [reply python/tests/core/test_statistics_engine.py:34]: Yes, since `FeatureGroup.__init__` calls `VariableApi.__init__`.

File: python/tests/core/test_transformation_function_engine.py (hunk 3 above)
Comment (aversey) [reply python/tests/core/test_transformation_function_engine.py:32]: Yes, since `FeatureGroup.__init__` calls `VariableApi.__init__`.

File: python/hopsworks_common/core/job_api.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+            _client._project_id,
+            "jobs",
+            type.lower(),
+            "configuration",
+        ]
+
+        headers = {"content-type": "application/json"}
+        return _client._send_request("GET", path_params, headers=headers)
+
+    def _delete(self, job):
```
Comment (vatj) [inline python/hopsworks_common/core/job_api.py:151]: Why is this one hidden vs the others?

File: python/hopsworks_common/core/job_api.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vatj) [reply python/hopsworks_common/core/job_api.py:252]: attach/detach makes sense to me but I think we had a discussion during dev-ex about it being unintuitive for non native speaker. In any case I think we might need to aim for some kind of consistency with the rest of the library and decide a path forward. I think we had a similar discussion with tags for example. Let's raise it up during standup

File: python/hopsworks_common/flink_cluster.py [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+    @property
+    def state(self):
+        """State of the cluster"""
+        current_exec = self._get_execution(assert_running=False)
+        if current_exec is not None:
+            return current_exec.state
+        else:
+            return None
+
+    def get_url(self):
```
[the remark this answers is not in the document]
Comment (vatj) [reply python/hopsworks_common/flink_cluster.py:381]: dataclass ?

File: python/tests/core/test_transformation_function_engine.py [hunk 9]
```diff
@@ -432,6 +435,7 @@ def testFunction1(col1):
     def test_get_and_set_feature_statistics_statistics_required(self, mocker):
         feature_store_id = 99
         mocker.patch("hopsworks_common.client.get_instance")
+        mocker.patch("hsfs.client.get_instance")
```
Comment (vatj) [inline python/tests/core/test_transformation_function_engine.py:438]: Why the double mocking?

File: python/tests/core/test_statistics_engine.py [hunk 10]
```diff
@@ -134,6 +134,7 @@ def test_compute_and_save_statistics_feature_view_obj(self, mocker):
         feature_store_id = 99
 
         mocker.patch("hopsworks_common.client.get_instance")
+        mocker.patch("hsfs.client.get_instance")
```
Comment (vatj) [inline python/tests/core/test_statistics_engine.py:137]: Why the double mocking?

**Review by vatj — APPROVED** [review]

File: python/hopsworks_common/core/job_api.py (hunk 7 above)
Comment (aversey) [reply python/hopsworks_common/core/job_api.py:151]: I am not sure, but it was this way in the original code.

File: python/hopsworks_common/execution.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+        """Name of the job the execution belongs to"""
+        return self._job.name
+
+    @property
+    def job_type(self):
+        """Type of the job the execution belongs to"""
+        return self._job.job_type
+
+    @property
+    def state(self):
```
[the remark this answers is not in the document]
Comment (aversey) [reply python/hopsworks_common/execution.py:100]: Do you know what exactly it can be? There are `contants.JOBS.SUCCESS_STATES` and `constants.JOBS.ERROR_STATES`, but they don't include in-progress states.

[the remark this answers is not in the document]
Comment (aversey) [reply python/hopsworks_common/execution.py:100]: Ok, I have found this: https://github.com/logicalclocks/hopsworks/blob/29d2bfe8a122821c6df2a5755601326536af31b1/hopsworks-persistence/src/main/java/io/hops/hopsworks/persistence/entity/jobs/configuration/history/JobState.java#L49
